### ISSUE TITLE
Raspberry Pi build thread

### ISSUE THREAD

**Issue by AndrewBelt** [VCVRack/Rack#534]
This thread documents others' work in compiling Rack for RPi.

If you hit an obstacle while compiling Rack, post the steps (commands) you used and the error, and someone might be able to help you. Then, the next time someone attempts to build for RPi, they can pick up where you left off, and so on until someone builds it completely.

**Remember, surround code and teminal output with triple backticks ```.**

**Comment by Doray** [comment]
roundup from #516:
standard 'compile.mk' after '# Optimization
FLAGS += -O3 -march=nocona '
delivers '-march=nocona' and the advise to choose an architecture.
I adjusted 'compile.mk' to: 
'# Optimization
FLAGS += -O3 -march=armv8-a '
which renders: 'fatal error: xmmintrin.h: No such file or directory
 #include <xmmintrin.h>'
and there the cart lost its wheel, i will try other architectures and report only on improvement.
greets D.
( i am a low frequent internet user)

**Comment by AndrewBelt** [comment]
***Remember, surround code and teminal output with triple backticks ```.***

**Comment by AndrewBelt** [comment]
@Doray 
1. I've reformatted your post like I requested you to, so people can actually read it.
2. When posting build errors, please at least include the file and line number of the error.

**Comment by bontric** [comment]
@Doray 

Looking a bit into this i found some information [here](https://bugs.launchpad.net/kicad/+bug/1626278)

Hence you might try to change [src/engine.cpp:9](https://github.com/VCVRack/Rack/blob/v0.5/src/engine.cpp#L9):

```cpp
#include <xmmintrin.h>
```

to

```cpp
#include <unistd.h>
```

There might be some redefines required, but you can post the output of make here and we will see where this leads us.

Another note:
> Ahh i just want to make a pocket mainframe called snowwhite and the seven dwarfs, a Pi3 with 7 Pi zeroes it easy to transport. As of now it has 16 channels audio through HDMI.

As of now, I'm not sure whether using a raspberry pi cluster will improve your performance since Rack currently uses a single thread model. 
I also really want to see Rack running on a Pi one day but it seems to me like a long term goal. I can also see that having a raspberry pi cluster is pretty cool, but for now you should really consider getting a used laptop (e.g. You'll get a Thinkpad x230 < 200€/$ and it will certainly outperform the raspberry pi's  in a single-thread scenario)

**Comment by AndrewBelt** [comment]
In addition, you'll need to comment out
```cpp
	_MM_SET_FLUSH_ZERO_MODE(_MM_FLUSH_ZERO_ON);
```
in that file https://github.com/VCVRack/Rack/blob/v0.5/src/engine.cpp#L118

**Comment by AndrewBelt** [comment]
As for *why* people want to compile on RPi, it's probably to use in small form factors, like in a Eurorack module itself. However, since the Rack audio engine isn't separated from the Rack graphics engine, we'll need to wait for when I make a headless "libRack" library.

**Comment by Doray** [comment]
Hi i did as you (bontronic and AndrewBelt) suggested, which leads us to the next step:
"Rack-0.5 $ make
g++ -Iinclude -Idep/include -Idep/lib/libzip/include -g -O3 -march=native -ffast-math -fno-finite-math-only -Wall -Wextra -Wno-unused-parameter -DARCH_LIN -Wsuggest-override -std=c++11 -c -o build/src/engine.cpp.o src/engine.cpp
g++ -Iinclude -Idep/include -Idep/lib/libzip/include -g -O3 -march=native -ffast-math -fno-finite-math-only -Wall -Wextra -Wno-unused-parameter -DARCH_LIN -Wsuggest-override -std=c++11 -c -o build/src/settings.cpp.o src/settings.cpp
In file included from include/app.hpp:4:0,
                 from src/settings.cpp:2:
include/widgets.hpp:5:38: fatal error: ../ext/nanovg/src/nanovg.h: No such file or directory
 #include "../ext/nanovg/src/nanovg.h"
                                      ^"
nonavg is an antialiasing lib for openGL perhaps Cairo will suffice ?
Note#2
i did set the ```-march=native``` in ```compile.mk:L9```

Note#3: A needed step i encountered is downloading Jansson which is :
"Jansson_ is a C library for encoding, decoding and manipulating JSON
data."
one can obtain it through:"http://www.digip.org/jansson/"
and read the docs at:"https://jansson.readthedocs.io/en/2.10/"

@AndrewBelt side note as to why. Please do not consider to make a libRack, because the whole fun of Rackvcv is the visual link.
I would rather have a bunch of Pi zeroes with small screens and one rackmodule on that screen interlinked in a big mess of wires then a faceless lib. There are a lot of these out there. Second point is if you did make a library out of it then a user would blame the interface computer for hampering not the one where the lib is running,
To sum it up i consider the power of this endeavor in the combination of visual interaction and audio output.
D.

**Comment by AndrewBelt** [comment]
```
#include "../ext/nanovg/src/nanovg.h"
```
This is an issue with not initializing your `git submodule`s. Instructions in the README. You should learn to compile Rack for x86_64 before attempting RPi.

Lol, no, you can't rip out nanovg and drop in cairo.

Yes, I will make a libRack library in the future. The Rack application will use libRack as the plugin loader and audio engine, while its job will be rendering the UI.

**Comment by Doray** [comment]
Well this is as far as i can get, i will start learning about submodules.

**Comment by rbino** [comment]
I've just succesfully compiled Rack and Fundamental plugins on my Raspberry Pi 3 with Raspbian Stretch.

I just followed the normal build procedure in the README with the following changes applied

```
diff --git a/compile.mk b/compile.mk
index 0ebc150..929a7bb 100644
--- a/compile.mk
+++ b/compile.mk
@@ -6,7 +6,7 @@ endif
 FLAGS += -MMD
 FLAGS += -g
 # Optimization
-FLAGS += -O3 -march=nocona -ffast-math -fno-finite-math-only
+FLAGS += -O3 -march=armv6 -mfloat-abi=hard -mfpu=vfp -ffast-math -fno-finite-math-only
 FLAGS += -Wall -Wextra -Wno-unused-parameter
 ifneq ($(ARCH), mac)
 CXXFLAGS += -Wsuggest-override
diff --git a/src/engine.cpp b/src/engine.cpp
index d7097e4..3859eca 100644
--- a/src/engine.cpp
+++ b/src/engine.cpp
@@ -6,7 +6,7 @@
 #include <algorithm>
 #include <chrono>
 #include <thread>
-#include <xmmintrin.h>
+#include <unistd.h>
 
 #include "engine.hpp"
 #include "util.hpp"
@@ -115,7 +115,7 @@ static void engineStep() {
 static void engineRun() {
        // Set CPU to denormals-are-zero mode
        // http://carlh.net/plugins/denormals.php
-       _MM_SET_FLUSH_ZERO_MODE(_MM_FLUSH_ZERO_ON);
+       //_MM_SET_FLUSH_ZERO_MODE(_MM_FLUSH_ZERO_ON);
 
        // Every time the engine waits and locks a mutex, it steps this many frames
        const int mutexSteps = 64;
```

I've done everything via ssh since I'm away from home but using X11 forwarding I verified that Rack starts and shows the main interface. I can't be sure it is producing audio but I don't see strange messaages in the console output.

I will check that everything is working correctly when I return home. I'm also not 100% sure about some optimization flags, when I'm home I will try some different combinations (e.g. `-vfpu=neon`). Do you have some suggested benchmarks to see if the flags make some difference?

**Comment by Doray** [comment]
Indeed it is working.
After installing the ```-dev``` version of ```libx11-dev, glew-dev, GLFW3, doxygen, libxinerama-dev, libxcursor-dev, asound-dev, libgtk2.0-cil-dev-2.12.40-2 and libgtk2.0-dev-2.24.31-2```
after ```make dep``` one types ```make run``` which returns an empty VCVRack with an enormous latency, powershortage warning (yellow lightning) and overheating warning(red thermometer), which is not workable.
So yes it works but it is not practical.
I want to thank AndrewBelt,Bontric and rbino, alas back to my Powermac.
Regards Doray

**Comment by rbino** [comment]
I can confirm that it's currently unusable on a Raspberry Pi 3, I've tried it with a lightweight WM (i3), I've modified the build flags to use neon FPU and applied all the performance enhancements described in #441 but Rack can barely run with just the Audio Interface, if a VCO is added it crashes.

**Comment by shreeswifty** [comment]
yes, even overclocked to 1.5, with a pisound. no love.

**Comment by pronvit** [comment]
I got VCVRack (git master) to work on RPi 3 +64bit OS without major difficulties. It’s usable if you add several modules only. The bad news is that the code is very very terribly unoptimised, it can do much much better, but I don’t have enough time at the moment to fix all the plugins. I’ll do what I can, although I’m not sure yet it’s worth it in general.

**Comment by pronvit** [comment]
So here's what I managed to achieve http://github.com/mi-rack/Rack
